### ISSUE TITLE
problem with stateUpdater(oldValue)

### ISSUE THREAD

**Issue by ryan-walsh-forte** [47ng/next-usequerystate#345]
I think https://github.com/47ng/next-usequerystate/blob/3cbeb3570ca36c948d15250d7876ff118a413b15/src/lib/app/useQueryState.ts#L238-L267 might be wrong in cases where `isUpdaterFunction` is true.

I think there is a race condition.

I'll let you know if I have clearer hints soon.

**Comment by franky47** [comment]
Yes, please review the `feat/v2` branch before going further though, it contains a completely different implementation.m that should solve some of these issues.

**Comment by ryan-walsh-forte** [comment]
Thanks.

**Comment by franky47** [comment]
There was indeed a race condition, even with the refactored code for 1.8.0: the old value being fed to the state updater function was being computed on the current URLSearchParams.

Because of the deferred/throttled update queue, if there were two state updating functions called before the queue was flushed, the first would have worked fine, but the second would have processed a stale base state, as the URL wouldn't have been updated yet with the result of the first call.

This was solved in 1.8.0-beta.15 by using the [Latest Ref pattern](https://epicreact.dev/the-latest-ref-pattern-in-react/), to feed state updater functions the last known state to be applied at the **next** render. This way, we can mix and match `useQueryState` and `useQueryStates` with overlapping keys without issues. A good side-effect of this change is that we no longer need to run the parsers on state updates, so it should be more performant.

The [crosslink demo](https://next-usequerystate-preview.vercel.app/demos/crosslink) showcases this behaviour ([page source](https://github.com/47ng/next-usequerystate/blob/next/src/app/demos/crosslink/page.tsx)).

Thanks for your report!

**Comment by ryan-walsh-forte** [comment]
Thanks for this helpful followup!

At the time, I ended up not digging into it further when I discovered that most of our app was using an existing home-built equivalent of useQueryState.

So I switched to using that and added a comment (reminder to myself) to eventually migrate us away from that home-built option since I'd prefer us to rely on libraries for this sort of thing since libraries will be more battle-hardened.

I've now subscribed to releases of useQueryState so will be notified.

Thanks!
